### ISSUE TITLE
Full‑pipeline regression test + readability overhaul + repo re‑layout + type‑hint cleanup

### ISSUE THREAD

**Issue by cristian1928** [cristian1928/Online-Adaptive-Deep-Residual-Neural-Network#15]
This umbrella issue brings the codebase to production shape by adding a
deterministic end‑to‑end test, rewriting unreadable sections, standardising the
directory structure, and eliminating leftover `# type: ignore[…]` pragmas.

---

### Goal  
Running the network with **exactly** the JSON configuration below must:

1. execute **all the way through** – including plotting and animation
   (use `matplotlib.use("Agg")` in CI);
2. emit a **tracking‑error norm** of **`4.657555165934869`** as produced by
   `DataManager.compute_tracking_error()`;
3. exit with status 0.

<details>
<summary>Fixture configuration</summary>

```json
{
  "final_time": 30,
  "time_step_delta": 0.001,

  "seed": 0,
  "num_states": 3,
  "control_size": 3,

  "target_initial_conditions": [40, 9, 2],
  "dynamics_type": "trophic_dynamics",

  "ID": "Residual Neural Network",
  "output_size": 3,
  "num_blocks": 4,
  "num_layers": 2,
  "num_neurons": 2,
  "inner_activation": "swish",
  "output_activation": "tanh",
  "shortcut_activation": "swish",

  "minimum_learning_rate": 0.01,
  "initial_learning_rate": 1,
  "maximum_learning_rate": 8,

  "weight_bounds": 4,

  "k1": 1
}
```
</details>

> **Note:** This integration test can take several minutes on slower hardware;
> CI timeout is set to **15 minutes**.

### Tasks  

| Step | Description |
|------|-------------|
| **1.1** | Add `tests/integration/test_tracking_error.py`. |
| **1.2** | Provide or wrap a public helper `run_pipeline(config: dict, headless: bool = True)` that returns the error norm. |
| **1.3** | Inside the test, load the fixture, call `run_pipeline`, then assert <br>`abs(error_norm - 4.657555165934869) < 1e-6`. |
| **1.4** | Mark the test `@pytest.mark.integration` and allow long runtime (`-k "integration" -m slow`). |

---

## 2 · Readability refactor (without breaking style gates)

Although the code passes Black/Ruff/Mypy, several sections are **painfully
opaque**.  Example:

```python
def train_step(self, step: int, loss: np.ndarray) -> np.ndarray:
    (
        _,
        neural_network_output,
        activated_layers_blocks,
        unactivated_layers_blocks,
        transposed_weights_blocks,
    ) = self._run_forward_pass(step)
    total_gradient: np.ndarray = self._run_backward_pass(
        activated_layers_blocks, unactivated_layers_blocks, transposed_weights_blocks
    )
    self.neural_network_gradient_wrt_weights = total_gradient
    self.update_neural_network_weights(step, loss)
    self.update_learning_rate(step)
    return neural_network_output
```

…and similar patterns in `forward_raw`, `jacobian_raw`, and
`update_learning_rate`. These inline tuple‑unpackings, deeply nested calls,
and anonymous inner functions seriously hamper comprehension.

### Refactor guidelines  

* **Extract helpers** – e.g. `_get_total_gradient(...)` or
  `self._forward_pass(step).output`.  
* **Rename variables** for intent (`loss_gradient` vs. `loss`).  
* **Zero docstrings/meta comments** – rely on expressive names + type hints.  
* Maintain existing style gates (Black, Ruff, Mypy).
* Check for redundancies and consolidate if possible, without breaking existing tests. 
* Ensure that computations are not being repeated erroneously, and fix if this is occurring. 

### Deliverables  

* Readability diff approved by ≥ 2 reviewers.  
* No behaviour change; all tests still pass.

---

## 3 · Professional repo re‑layout

Move to a modern **src‑layout**:

```
.
├── src/
│   └── rnn_sim/
│       ├── core/            # neural nets, dynamics, math utils
│       ├── io/              # config I/O, data manager
│       ├── viz/             # plotting / animation
│       └── apps/            # CLI entry‑points (main.py, train.py)
├── tests/
└── pyproject.toml
```

* Update dotted‑path imports.  
* Provide `python -m rnn_sim.apps.main --config cfg.json` entry‑point.  
* Relocate datasets/fixtures to `data/` or `tests/data/`.

---

## 4 · Eliminate remaining `# type: ignore[...]`

* Search for ignore pragmas and either tighten the types, refactor code, or
  use narrow `typing.cast`.  
* Final `grep -R "# type: ignore" src/` must return **0** lines.

---

## 5 · Acceptance checklist

- [ ] Integration test passes with error norm `4.657555165934869 ± 1e‑6`.  
- [ ] Readability improvements merged; sample functions now short, intent‑revealing, and unit‑tested.  
- [ ] New src‑layout builds & runs; imports resolve.  
- [ ] No `# type: ignore` remain.  
- [ ] Black, Ruff, Mypy & all tests green; CI completes within 15 min.  

---

### 📈 Outcome

*Deterministic pipeline test + human‑readable code + clean layout +
zero ignore‑pragmas = a maintainable, production‑ready repository.*
